### PR TITLE
Improvements to sampler package

### PR DESCRIPTION
**Pull request by jotok**
* Standardize field names in CompactSampler
* Update Javadocs
* Add ability to validate the heap property in CompactSampler
* Move common methods to the IStreamSampler interface
* Add `@Data` annotation to Weighted class and remove unused methods
* Improve package test coverage
* Set fields in CompactSamplerState to private

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by jotok** [comment]
The initial commit failed to build due to a javadoc error (namely, using `<p/>` instead of `<p>` in a javadoc comment). Fixed and amended commit.

**Comment by jotok** [comment]
Rebased.

File: Java/core/src/main/java/com/amazon/randomcutforest/sampler/CompactSampler.java [hunk 1]
```diff
@@ -163,58 +197,75 @@ public static CompactSampler uniformSampler(int sampleSize, long seed, boolean s
     public Optional<Float> acceptSample(long sequenceIndex) {
         evictedPoint = null;
         float weight = computeWeight(sequenceIndex);
-        entriesSeen++;
         if (size < capacity) {
```
Comment (sudiptoguha) [inline Java/core/src/main/java/com/amazon/randomcutforest/sampler/CompactSampler.java:200]: Do we want to store the information that acceptSample was called and some weight generated, even if nothing was evicted?

File: Java/core/src/main/java/com/amazon/randomcutforest/sampler/CompactSampler.java [hunk 2]
```diff
@@ -163,58 +197,75 @@ public static CompactSampler uniformSampler(int sampleSize, long seed, boolean s
...
+            swapDown(i, validate);
         }
     }
 
     @Override
-    public void addSample(Integer pointRef, float weight, long seqNUm) {
-        checkState(size < capacity, " sampler full");
-        weightArray[size] = weight;
-        referenceArray[size] = pointRef;
+    public void addSample(Integer pointIndex, float weight, long sequenceIndex) {
```
Comment (sudiptoguha) [inline Java/core/src/main/java/com/amazon/randomcutforest/sampler/CompactSampler.java:258]: Do we want to check if a acceptSample was called (getEvictedPoint)? And reset that state? That would be true if size = capacity - 1.

Should we check for weights generated earlier in an acceptSample even if nothing was evicted? Probably not. But this does leaf the feature that we can artificially change the weight of a an element that is sampled to have different duration in the sampler.

**Review by sudiptoguha — APPROVED** [review]

**Pull request merged**